File: cplusplus/design_pattern/composite_pattern/Composite_trans.cpp
#include "stdafx.h"
#include "Composite_trans.h"


Composite_trans::Composite_trans()
{
}


Composite_trans::~Composite_trans()
{
}

void Composite_trans::addChild(std::shared_ptr<IComponent_trans> child)
{
    myCollection.push_back(child);
}

void Composite_trans::deleteChild(std::shared_ptr<IComponent_trans> child)
{
    auto position = std::find(myCollection.begin(), myCollection.end(), child);
    if (position == myCollection.end())
        return;
    myCollection.erase(position);
}

void Composite_trans::loopChilds()
{
    displayClassName();
    std::for_each(myCollection.begin(), myCollection.end(), [](std::shared_ptr<IComponent_trans> const & it)->void {it->loopChilds(); });
}

void Composite_trans::displayClassName()
{
    std::cout << "Composite_trans::displayClassName()" << std::endl;
}

File: Java/pattern/composite_pattern/src/com/pattern/Composite_trans.java
package com.pattern;

import java.util.ArrayList;

public class Composite_trans implements IComponent_trans {
	private ArrayList<IComponent_trans> myCollection = new ArrayList<IComponent_trans>();

	@Override
	public void addChild(IComponent_trans child) {
		myCollection.add(child);
	}

	@Override
	public void deleteChild(IComponent_trans child) {
		if (myCollection.contains(child))
			myCollection.remove(child);
	}

	@Override
	public void loopChild() {
		int collectionSize = myCollection.size();
		for (int i = 0; i < collectionSize; ++i) {
			myCollection.get(i).displayClassName();
			if (myCollection.get(i) instanceof com.pattern.Composite_trans) {
				myCollection.get(i).loopChild();
			}
		}

	}

	@Override
	public void displayClassName() {
		Class<?> enclosingClass = getClass().getEnclosingClass();
		if (enclosingClass != null) {
		  System.out.println(enclosingClass.getName());
		} else {
		  System.out.println(getClass().getName());
		}
	}
}
